mpc85xx: add support for Edge-core ECW7210-L

zeroping opened the pull request:
Hardware
========
- CPU: Freescale P1020 PowerPC
- RAM: 256 M DDR3 Nanya 2x NT5CB64M16FP-DH (128 each)
- NOR: 32MiB MXIC S134790MX29GL256FHT2I-90Q (u-boot)
- NAND: 512MiB Micron 29F4G08ABADA (system images)
- Eth0: BCM54612E PoE
- Eth1: BCM54612E
- 2.4GHz WiFi: 14e4:4331 BCM4331 802.11a/b/g/n
- 5.0GHz WiFi: 14e4:43a2 BCM4360 802.11ac
- USB: 1x USB 2.0 (internal)
- LED: Power, WiFi, LAN ( all non-controllable)
- BTN: Reset (harware)

Serial is a Cisco-compatible RJ45 jack next to the ethernet ports.
115200-N-8 are the settings for OS and U-Boot.

Installation
============
1.  Power on the device while attached to the console port.

2.  Halt the U-Boot by pressing Enter when prompted.
2a. Boot loader may be configured to always autoboot (observed on a Tallac
    device). If this is the case, you need to corrupt the reads from the 
    NAND after u-boot has started, but before the kernel image is read. 
    It's U32, under the PoE board. Short pin 25 (GND) to any pin 29-32 at 
    just the right time, and u-boot should see the failed checksum and drop
    you to a console.

    If you're dealing with this, temporarily disable autoboot by setting
    bootcmd to something useless:
  > setenv bootcmd echo "This u-boot has a hardcoded autoboot"
  > saveenv

    When you're done, you might consider leaving yourself a way in. I use a
    USB key plugged in to cause the boot loader to stop autoboot:
  > setenv bootcmd run boot_openwrt
  > setenv boot_openwrt "echo Autoboot broken on this U-boot. Insert a USB 
    key to abort boot; usb start; if usb storage; then echo aborting\
    autoboot; else echo autobooting; nboot 0x1000000 0 0 && \
    bootm 0x1000000 ; fi"

3.  Set the correct bootcmd for booting OpenWRT:
  > setenv boot_openwrt "setenv bootargs console=ttyS0,115200; \
    nboot 0x1000000 0 0 && bootm 0x1000000"
  > setenv bootcmd run boot_openwrt
  > saveenv

4.  Rename OpenWRT initramfs image to 'kernel.bin' and place it in a
    TFTP server root-directory.

5.  Configure ip addresses:
  > setenv ipaddr 192.168.0.250 (a static ip on your network)
  > setenv serverip 192.168.0.42 (your tftp server's IP)

6.  Set up the u-boot environment variables for the Ethernet MAC addresses:
    (br-lan won't work if there's not a mac address set for eth0)
  > setenv ethaddr  70:72:CF:00:12:10
  > setenv eth2addr 70:72:CF:00:12:02
  > saveenv

4.  Boot OpenWRT initramfs image with:
  > setenv bootargs console=ttyS0; tftpboot kernel.bin; bootm

6.  (Optional)
    Make a Backup of the NAND flash in case you ever want to go back to the
    vendor firmware. We will be reformatting it entirely, so you may want to
    record the layout shown during stock kernel boot.

8.  Transfer OpenWRT sysupgrade image to the device via SCP and install it
    with:
  > sysupgrade -n \<openwrt-image-file\>


Signed-off-by: Scott Gilliland <zeroping@gmail.com>

zeroping committed:
mpc85xx: add support for Edgecore ECW7210-L

Hardware
========
- CPU: Freescale P1020 PowerPC
- RAM: 256 M DDR3 Nanya 2x NT5CB64M16FP-DH (128 each)
- NOR: 32MiB MXIC S134790MX29GL256FHT2I-90Q (u-boot)
- NAND: 512MiB Micron 29F4G08ABADA (system images)
- Eth0: BCM54612E PoE
- Eth1: BCM54612E
- 2.4GHz WiFi: 14e4:4331 BCM4331 802.11a/b/g/n
- 5.0GHz WiFi: 14e4:43a2 BCM4360 802.11ac
- USB: 1x USB 2.0 (internal)
- LED: Power (non-controllable), WiFi (non-controllable), LAN (non-controllable)
- BTN: Reset (harware)

Serial is a Cisco-compatible RJ45 jack next to the ethernet ports.
115200-N-8 are the settings for OS and U-Boot.

Installation
============
1.  Power on the device while attached to the console port.

2.  Halt the U-Boot by pressing Enter when prompted.
2a. Bootloader may be configured to always autoboot (observed on a Tallac
    device). If this is the case, you need to corrupt the reads from the NAND
    after u-boot has started, but before the kernel image is read. It's the
    U32, under the PoE board. Short pin 25 (GND) to any pin 29-32 at just the
    right time, and u-boot should see the failed checksum and drop you to a
    console.

    If you're dealing with this, temporarily disable autoboot by setting
    bootcmd to something useless:
  > setenv bootcmd echo "This u-boot has a hardcoded autoboot"
  > saveenv

    When you're done, you might consider leaving yourself a way in. I use a
    USB key plugged in to cause the bootloader to stop autoboot:
  > setenv bootcmd run boot_openwrt
  > setenv boot_openwrt "echo Autoboot broken on this U-boot. Insert a USB key\
    to abort boot; usb start; if usb storage; then echo aborting autoboot;\
    else echo autobooting; nboot 0x1000000 0 0 && bootm 0x1000000 ; fi"

3.  Record the current environment with printenv. You may specifically need
    the ethaddr and eth1addr values, as these appear uniquie per device and
    are used to set the MAC addresses of the ethernet interfaces.

4.  Set the correct bootcmd for booting OpenWRT:
  > setenv boot_openwrt "setenv bootargs console=ttyS0,115200; \
    nboot 0x1000000 0 0 && bootm 0x1000000"
  > setenv bootcmd run boot_openwrt
  > saveenv

5.  Rename OpenWRT initramfs image to 'kernel.bin' and place it in a
    TFTP server root-directory.

6.  Configure ip addresses:
  > setenv ipaddr 192.168.0.250 (a static ip on your network)
  > setenv serverip 192.168.0.42 (your tftp server's IP)

7.  Check the u-boot enviroment variables for the ethernet MAC addresses.
    These must be something valid and unique, ideally the values from
    step 3. It's not clear why eth2addr needs to be set, but it functions
    to set the MAC for eth0.
    (br-lan won't work if there's not a mac address set for eth0)
  > setenv ethaddr  70:72:CF:B3:BC:00 (should already be set)
  > setenv eth2addr 70:72:CF:B3:BC:10 (use value from eth1addr)
  > saveenv

8.  Boot OpenWRT initramfs image with:
  > tftpboot kernel.bin; bootm

9.  (Optional)
    Make a Backup of the NAND flash in case you ever want to go back to the
    vendor firmware. We will be reformatting it entirely, so you may want to
    record the layout shown during stock kernel boot. NOR flash (containing
    u-boot) will be untouched.

10. Transfer OpenWRT sysupgrade image to the device via SCP and install it
    with:
  > sysupgrade -n <openwrt-image-file>

Signed-off-by: Scott Gilliland <zeroping@gmail.com>

adrianschmutzler commented:
Hi, please use a commit title according to https://openwrt.org/submitting-patches, e.g.
"mpc85xx: add support for Edge-Core ECW7210-L"

adrianschmutzler commented on target/linux/mpc85xx/base-files/lib/upgrade/platform.sh:
```diff
@@ -14,7 +14,8 @@ platform_do_upgrade() {
 	ocedo,panda|\
-	sophos,red-15w-rev1)
+	sophos,red-15w-rev1|\
+	edgecore,ecw7210)
```
Please take care of alphabetic sorting. Despite, the device should be named edgecore,ecw7210-l (or edge-core,ecw7210-l)

adrianschmutzler commented on target/linux/mpc85xx/files/arch/powerpc/boot/dts/ecw7210.dts:
```diff
@@ -0,0 +1,304 @@
+ * option) any later version.
+ */
+
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
```
You arleady have a license above. Remove the above version and stick to the license identifier here.

adrianschmutzler commented on target/linux/mpc85xx/files/arch/powerpc/boot/dts/ecw7210.dts:
```diff
@@ -0,0 +1,304 @@
+/include/ "fsl/p1020si-pre.dtsi"
+/ {
+	model = "Edge-Core ECW7210-L";
+	compatible = "edgecore,ecw7210";
```
So, again either "edge-core,ecw7210-l" or "edgecore,ecw7210-l". Adjust file name accordingly "ecw7210-l.dts"

adrianschmutzler commented on target/linux/mpc85xx/files/arch/powerpc/boot/dts/ecw7210.dts:
```diff
@@ -0,0 +1,304 @@
+		#size-cells = <0x1>;
+		#address-cells = <0x2>;
+		bus-frequency = <24999999>;
+		nand@1,0 {
```
Please add a space before that node and between the child nodes.

adrianschmutzler commented on target/linux/mpc85xx/files/arch/powerpc/boot/dts/ecw7210.dts:
```diff
@@ -0,0 +1,304 @@
+			};
+			
+			//label = "ubi";
+			//compatible = "ubi,volume";
```
Remove comments from testing.

adrianschmutzler commented on target/linux/mpc85xx/files/arch/powerpc/boot/dts/ecw7210.dts:
```diff
@@ -0,0 +1,304 @@
+		};
+	};
+
+	board_soc: soc: soc@ffe00000 {
```
Decide for one label, if you need it. I do not think you use any of those labels.

blocktrron replied:
@adrianschmutzler due to the way the upstream tree works, you need at least the `soc` label (same goes for the localbus `lbc` label).

The `board_lbc` as well as the `board_soc` here can be dropped, as they are not used (anymore?).

zeroping replied:
Yep, this is on my to-do list under "fix everything in the .dts because it's awful right now". I carried some cruft over from the stock device tree while fixing PCIe IRQ's. I don't fully understand how I managed to get it working, but that's not a good excuse to leave the .dts untidy.

zeroping replied:
I've dropped the unnecessary ones. Only leaving `lbc`, `soc`, `enetx` and `pcix` to work with the included p1020 dtsi's, as blocktrron mentioned.

adrianschmutzler commented on target/linux/mpc85xx/files/arch/powerpc/boot/dts/ecw7210.dts:
```diff
@@ -0,0 +1,304 @@
+			status = "okay";
+		};
+
+                mdio@25000 {
```
Indent!

adrianschmutzler commented on target/linux/mpc85xx/files/arch/powerpc/boot/dts/ecw7210.dts:
```diff
@@ -0,0 +1,304 @@
+
+                mdio@25000 {
+			compatible = "fsl,etsec2-tbi";
+			tbi_phy1: tbi-phy@12 {
```
Add empty line before.

adrianschmutzler commented on target/linux/mpc85xx/files/arch/powerpc/boot/dts/ecw7210.dts:
```diff
@@ -0,0 +1,304 @@
+
+
+		
+                eth0: enet2: ethernet@b2000 {
```
- Remove multiple empty lines
- Remove multi-label
- Fix indent (also below, I won't comment again about that).

adrianschmutzler commented on target/linux/mpc85xx/files/arch/powerpc/boot/dts/ecw7210.dts:
```diff
@@ -0,0 +1,304 @@
+			model = "eTSEC";
+			phy-handle = <&phy1>;
+			tbi-handle = <&tbi_phy1>;
+			local-mac-address = [70 72 cf 42 42 03];		
```
You are setting a static MAC address here??

zeroping replied:
It looks like I need some kind of MAC here, even if it's made up. U-boot will replace it if everything works correctly (if ethaddr and ethaddr2 are set). I had considered 00:00:00:00:00:00, but if u-boot doesn't set it, we end up without a MAC on eth0, and that causes br-lan not to come up. 

70:72:CF is Edge-core's OUI, and I figured 42:42:xx would be as good of a MAC as any.

I figured that we'd be better off with a made-up MAC than an inaccessible device, but anyone working on this will undoubtedly have a console. I'll defer to your judgement - just let me know if you want these two MACs taken out.

adrianschmutzler replied:
You have to _read_ the MAC address from some location. Have look at how this is done for other devices. I've made an incomplete tutorial here: https://openwrt.org/docs/guide-developer/mac.address

zeroping replied:
In the stock system, the MAC addresses appear to come from u-boot environment variables. It's not actively read by the OS from anywhere, as on other systems - it appears to be copied into the stock DTS by u-boot, on top of whatever value is entered into those `local-mac-address` properties.

Now, I could be wrong. Changing those u-boot environment variables seemed to change the MAC's in the DTS got handed to the kernel, but it's also possible that those MACs can be found somewhere in the NOR or NAND flash. I haven't done that search, and need to. I can also confirm that `ethaddr` and `ethaddr2` are set with the correct MAC's on a stock system, and that they're not junk left in the environment from the stock setup.

I'll confirm that I have to have the `local-mac-address` property for the u-boot-DTS-MAC-copying to work. I can just set it to all 00's, which effectively disables the interface if it u-boot doesn't set it. (the eTSEC driver appears to disallow `ifconfig up` on the interface if there's no MAC address). Would that be the best choice if the property has to be there?

zeroping replied:
The MAC's are no longer set in the dts. This leaves a non-functional router if the u-boot environment gets reset for some reason. The stock OS really does rely on the MAC addresses being set by u-boot, but they work around u-boot environment issues by aggressively detecting and re initializing the u-boot environment if it's changed (including re-locking u-boot's autoboot!).

One note: the stock OS uses `ethaddr` and eth1addr` to set MACs. Under Openwrt, we need to instead use `ethaddr` and `eth2addr` to set the MACs. I suspect this is because of the way u-boot is searching through the device tree. I've clarified info about setting those environment variables to the commit message.

adrianschmutzler commented on target/linux/mpc85xx/files/arch/powerpc/boot/dts/ecw7210.dts:
```diff
@@ -0,0 +1,304 @@
+		};
+		
+	
+
```
Multiple empty lines...

adrianschmutzler commented on target/linux/mpc85xx/files/arch/powerpc/boot/dts/ecw7210.dts:
```diff
@@ -0,0 +1,304 @@
+		};
+	};
+	
+
```
Remove empty lines.

adrianschmutzler commented on target/linux/mpc85xx/files/arch/powerpc/boot/dts/ecw7210.dts:
```diff
@@ -0,0 +1,304 @@
+/include/ "fsl/p1020si-post.dtsi"
+
+
+
```
also here, please fix all of those cases, I won't comment again.

adrianschmutzler commented on target/linux/mpc85xx/files/arch/powerpc/boot/dts/ecw7210.dts:
```diff
@@ -0,0 +1,304 @@
+
+
+
+/ {
```
Why do you have a second root node here? Please tidy up your file before submitting it.

zeroping replied:
Some of this is left-over from a week's worth of hacking. Don't get me wrong, I have some cleanup to do, and may have rushed this a bit to get it in before work started back.

That said, some of this second node may be necessary: I'd prefer to include fsl/p1020si-post.dtsi, which includes fsl/pq3-duart-0.dtsi, which then sets things like: 
`clock-frequency = <0>; `

My understanding is that u-boot is supposed to set that to a real value, but that doesn't happen with this device's u-boot. So, if I understand correctly, my options are:

1. Copy in fsl/p1020si-post.dtsi minus the include for fsl/pq3-duart-0.dtsi, and copy in aversion of fsl/pq3-duart-0.dtsi with the frequency set correctly.
2. Include fsl/p1020si-post.dtsi and then override it with a second node like this one to set the frequency.

I think approach 2 is cleaner, and looks like how a issue was handled in tl-wdr4900-v1.dts.

I might not be understanding correctly, and I will test it more when I do the rest of the .dts cleanup. In the least, I think I should narrow it down the the minimum changes needed after include of fsl/p1020si-post.dtsi (likely just the frequencies).

Let me know if that doesn't sound like a good approach.

zeroping replied:
I did a _lot_ of experimenting, and got the .dts to be much cleaner. Thanks for pushing me to do so - see if it looks OK now.

adrianschmutzler commented on target/linux/mpc85xx/files/arch/powerpc/platforms/85xx/ecw7210.c:
```diff
@@ -0,0 +1,95 @@
+
+	fsl_pci_assign_primary();
+
+        printk(KERN_INFO "ECW7210 board from Edge-core\n");
```
As before, please take care of your indent before submitting (applies to whole file as well).

adrianschmutzler commented on target/linux/mpc85xx/image/Makefile:
```diff
@@ -126,6 +126,23 @@ define Device/ocedo_panda
 endef
 TARGET_DEVICES += ocedo_panda
 
+define Device/ecw7210
```
Look for guidance on proper name above, like compatible, just with "_" instead of comma.

adrianschmutzler commented on target/linux/mpc85xx/image/Makefile:
```diff
@@ -126,6 +126,23 @@ define Device/ocedo_panda
 TARGET_DEVICES += ocedo_panda
 
+define Device/ecw7210
+  DEVICE_TITLE := Edge-core ECW7210
```
Please use the new syntax introduced several months ago.

adrianschmutzler commented on target/linux/mpc85xx/image/Makefile:
```diff
@@ -126,6 +126,23 @@ define Device/ocedo_panda
+  PAGESIZE := 2048
+  SUBPAGESIZE := 512
+  BLOCKSIZE := 128k
+  SUPPORTED_DEVICES := edgecore,ecw7210
```
That's not needed if you choose names correctly.

adrianschmutzler commented on target/linux/mpc85xx/image/Makefile:
```diff
@@ -126,6 +126,23 @@ define Device/ocedo_panda
+  IMAGES := fdt.bin sysupgrade.bin
+  IMAGE/fdt.bin := append-dtb
+  IMAGE/sysupgrade.bin := sysupgrade-tar | append-metadata
+
```
Remove empty line.

zeroping commented:
It will take me a little bit to get to all of these, but I'll start here:
> Hi, please use a commit title according to https://openwrt.org/submitting-patches, e.g.
> "mpc85xx: add support for Edge-Core ECW7210-L"

I'm a little confused here. Isn't that the exact title I started with, and now have? I know I mucked things up when I tried to add "label:target/mpc85xx" myself, but I changed that back. Let me know if I still need to make changes to the title.

I'll address my other questions in-line with your comments.

Thanks for your patience with me here - I've clearly made a mess of the .dts.

adrianschmutzler commented:
> I'm a little confused here. Isn't that the exact title I started with, and now have? I know I mucked things up when I tried to add "label:target/mpc85xx" myself, but I changed that back. Let me know if I still need to make changes to the title.

That's the title of the Pull Request, which is relatively irrelevant. I'm talking about the title of the commit, which is still "Add the Edge-core ECW7210-L under the mpc85xx/p1020 subtarget.". At the end, only the information in the commit will be preserved.

blocktrron commented on target/linux/mpc85xx/files/arch/powerpc/platforms/85xx/ecw7210.c:
```diff
@@ -0,0 +1,95 @@
+	if (of_machine_is_compatible("edgecore,ecw7210"))
+		return 1;
+        //added so that we have a chance of working with the stock DTB found in NOR flash
+        if (of_machine_is_compatible("fsl,P1020RDB"))
```
Is any comatibility given with a recent OpenWrt kernel and the vendor tree? If not, i would drop this here.

zeroping replied:
It at least thinks about booting and you get a serial console. Without this, there's no chance. The vendor .dtb sits in it's own partition in NOR flash, so I had it in there for that exact combination (new kernel, old .dtb). I can pull it out though. That was another thing left over from debugging, and probably not really needed now.

blocktrron replied:
Yup, please drop this.

blocktrron commented on target/linux/mpc85xx/image/Makefile:
```diff
@@ -126,6 +126,23 @@ define Device/ocedo_panda
+  DEVICE_TITLE := Edge-core ECW7210
+  DEVICE_PACKAGES := uboot-envtools
+  # this u-boot supports fit images, but has issues with gzipped kernels
+  # larger than about 4.3 MB. Thus, we use an uncompressed kernel.
```
Could you elaborate on the issues you were experiencing? Maybe we can fix the underlying issue instead.

zeroping replied:
Regarding the old-style DEVICE_TITLE format, it's because I gave up on 19 for a bit, and went back to 17, got the device at least booting there, and then brought my changes forward to 18, then 19. The result is that I didn't pay attention when copying-and-pasting, and didn't see the new format.

As to the 4.3 MB size issue, I was getting `GUNZIP: uncompress, out-of-mem or overwrite error - must RESET board to recover`, see more [here](https://pastebin.com/7Ag5eG0a).  See also the similar-looking [mpc85xx: Use uncompressed kernel on the HiveAP-330](https://github.com/openwrt/openwrt/pull/2162). Based on my reading, it sounds like there's a compile-time option in the u-boot source that hard-caps the kernel image size, and this u-boot just happens to have it set kind of low. The good news is that we have tons of NAND flash, so uncompressed isn't much of an issue.

adrianschmutzler commented:
> Regarding the old-style DEVICE_TITLE format, it's because I gave up on 19 for a bit, and went back to 17, got the device at least booting there, and then brought my changes forward to 18, then 19.

Note that DEVICE_VENDOR/DEVICE_MODEL is only in master, so you will have to change to DEVICE_TITLE even if you backport to 19.07 locally.

adrianschmutzler commented on target/linux/mpc85xx/image/Makefile:
```diff
@@ -126,6 +126,23 @@ define Device/ocedo_panda
 TARGET_DEVICES += ocedo_panda
 
+define Device/edgecore_ecw7210-l
+  DEVICE_VENDOR := Edgecore
```
In DTS model property you call it "Edge-Core". Please be consistent.

zeroping replied:
Their website is edge-core.com, yet the company name is Edgecore Networks, and they use Edgecore in almost all of the text on their site. The logo itself is stylized as 'Edge-coreE', and that capitalization is definitely not going to be carried through Makefiles. I've decided to just use edgecore/Edgecore only, no hypen, no funny capitalization.

adrianschmutzler replied:
Okay, the change the "model" in DTS accordingly.

zeroping replied:
I don't understand quite what you mean.

My intent is to refer to the company as "Edgecore" when capitalization seems to be the norm, and "edgecore" when not. Likewise, my intent is to go with "ECW7210-L" for the model when it seems like capitalization seems to be the norm, and "ecw7210-l" when not. Thus, I've changed the Makefile to be `DEVICE_VENDOR := Edgecore` and `DEVICE_MODEL := ECW7210-L`. This seems to match the capitalization and hyphenation of the other mpc85xx targets.

If I've screwed something up with git(hub), and you're not seeing any of my recent changes, then that's a different problem.

adrianschmutzler replied:
Yes, and I want you to change the "model" string in your DTS file (!) from
`model = "Edge-Core ECW7210-L";`
to
`model = "Edgecore ECW7210-L";`
as well.

adrianschmutzler replied:
Ping.

Either change to
DEVICE_VENDOR := Edge-Core
here or put the following in the DTS:
model = "Edgecore ECW7210-L";

zeroping commented:
Alright, I think I've resolved all the issues from before:
- The .dts is _much_ cleaner now
- I've standardized on "edgecore" and "ECW7210-L" for naming
- I've confirmed how MACs are set in the stock OK, and we now do similar.

I hope that I haven't missed anything, but let me know what you see.
(also, the workflow where I squash everything into an amended commit for the PR is a bit new to me, so let me know if I've totally screwed that up.)

tmn505 commented:
I assume wifi is not supported by any driver, so would appreciate big fat warning about it in commit message.

zeroping commented:
> I assume wifi is not supported by any driver, so would appreciate big fat warning about it in commit message.

Oh, good point. I'd be happy to add that, but I could use some suggestions first. 

What's the right thing to do for a Broadcom BCM4331 and BCM4360? I believe the BCM4331 may be able to work at G-only speeds, and only 2.4 GHz, but I interpreted that as being so bad so as to be useless. I admit that I swapped Atheros cards into my devices.

Should I change the Makefile to build in any driver that works, no matter how badly? Or build the image with less in it, and let people install the wireless modules they want?

I can swap cards back in to test if needed.

tmn505 commented:
It would be good to test those cards. But I personally, wouldn't bother to add the drivers for them if they don't allow to use full potential of the hardware. Decision is up to You.

If You still have issues with assigning MAC addresses, You can use *aliases* node to let the U-Boot know in which node to write MAC to. So for example: `ethernet0 = &enet2` would write `ethaddr` to `enet2` node, `ethernet1 = &enet0` would write `eth1addr` to enet0 node. Might be that this U-Boot assigns MAC from `ethaddr1` instead of `eth1addr`.

adrianschmutzler commented:
In the meantime, kernel 5.4 has arrived, and kernel 4.14 has been dropped (for the target) ...

adschm commented:
Please update for kernel 5.4.

zeroping committed:
mpc85xx: add support for Edgecore ECW7210-L

Hardware
========
- CPU: Freescale P1020 PowerPC
- RAM: 256 M DDR3 Nanya 2x NT5CB64M16FP-DH (128 each)
- NOR: 32MiB MXIC S134790MX29GL256FHT2I-90Q (u-boot)
- NAND: 512MiB Micron 29F4G08ABADA (system images)
- Eth0: BCM54612E PoE
- Eth1: BCM54612E
- 2.4GHz WiFi: 14e4:4331 BCM4331 802.11a/b/g/n
- 5.0GHz WiFi: 14e4:43a2 BCM4360 802.11ac
- USB: 1x USB 2.0 (internal)
- LED: Power (non-controllable), WiFi (non-controllable), LAN (non-controllable)
- BTN: Reset (harware)

Serial is a Cisco-compatible RJ45 jack next to the ethernet ports.
115200-N-8 are the settings for OS and U-Boot.

Installation
============
1.  Power on the device while attached to the console port.

2.  Halt the U-Boot by pressing Enter when prompted.
2a. Bootloader may be configured to always autoboot (observed on a Tallac
    device). If this is the case, you need to corrupt the reads from the NAND
    after u-boot has started, but before the kernel image is read. It's the
    U32, under the PoE board. Short pin 25 (GND) to any pin 29-32 at just the
    right time, and u-boot should see the failed checksum and drop you to a
    console.

    If you're dealing with this, temporarily disable autoboot by setting
    bootcmd to something useless:
  > setenv bootcmd echo "This u-boot has a hardcoded autoboot"
  > saveenv

    When you're done, you might consider leaving yourself a way in. I use a
    USB key plugged in to cause the bootloader to stop autoboot:
  > setenv bootcmd run boot_openwrt
  > setenv boot_openwrt "echo Autoboot broken on this U-boot. Insert a USB key\
    to abort boot; usb start; if usb storage; then echo aborting autoboot;\
    else echo autobooting; nboot 0x1000000 0 0 && bootm 0x1000000 ; fi"

3.  Record the current environment with printenv. You may specifically need
    the ethaddr and eth1addr values, as these appear uniquie per device and
    are used to set the MAC addresses of the ethernet interfaces.

4.  Set the correct bootcmd for booting OpenWRT:
  > setenv boot_openwrt "setenv bootargs console=ttyS0,115200; \
    nboot 0x1000000 0 0 && bootm 0x1000000"
  > setenv bootcmd run boot_openwrt
  > saveenv

5.  Rename OpenWRT initramfs image to 'kernel.bin' and place it in a
    TFTP server root-directory.

6.  Configure ip addresses:
  > setenv ipaddr 192.168.0.250 (a static ip on your network)
  > setenv serverip 192.168.0.42 (your tftp server's IP)

7.  Check the u-boot enviroment variables for the ethernet MAC addresses.
    These must be something valid and unique, ideally the values from
    step 3. It's not clear why eth2addr needs to be set, but it functions
    to set the MAC for eth0.
    (br-lan won't work if there's not a mac address set for eth0)
  > setenv ethaddr  70:72:CF:B3:BC:00 (should already be set)
  > setenv eth2addr 70:72:CF:B3:BC:10 (use value from eth1addr)
  > saveenv

8.  Boot OpenWRT initramfs image with:
  > tftpboot kernel.bin; bootm

9.  (Optional)
    Make a Backup of the NAND flash in case you ever want to go back to the
    vendor firmware. We will be reformatting it entirely, so you may want to
    record the layout shown during stock kernel boot. NOR flash (containing
    u-boot) will be untouched.

10. Transfer OpenWRT sysupgrade image to the device via SCP and install it
    with:
  > sysupgrade -n <openwrt-image-file>

Signed-off-by: Scott Gilliland <zeroping@gmail.com>

zeroping committed:
Merge changes from 'edgecore-kernel5.4' for kernel 5.4 updates

zeroping commented:
The updates for 5.4 were pretty straightforward - this platform is just like every other p1020 platform. I have my freshly-built image up and running on my device here.

I've done what I can, but I may have botched my fork's git commit history - I notice that github is reporting that there are conflict with this pull request. Let me know what (if anything) I need to change, and I'll do my best.

zeroping commented:
Would you like me to just close this and open a new, cleaner PR, with a branch that is forked from OpenWRT's current master branch?

adschm commented:
No, you should rebase/squash this (git rebase -i ...) and force-push the result (git push -f ...) to _this_ PR.

adschm commented:
Still needs to be squashed into one commit.

CodeFetch commented:
@adschm Why don't you squash it with Github? Just curious...

adschm commented:
@CodeFetch 

1. Having to clean up myself will mean more work for me
2. The author's SoB should cover as much as possible, so he should do most of the changes
3. If I regularly clean up stuff like this, the author will never learn to do it properly, and thus I will have to do it every time in the future as well
4. I do not merge device support on this target, so I won't pick it anyway, but nobody else will unless it properly squashed

adschm commented:
Since there has been no interest in reviewing this for more than one year, and the PR would need additional work anyway, I'm closing this now.

Sorry, but I don't think waiting another year will help for anything here.
